### PR TITLE
[PIN-2875] Negate the possibility to creates two e-services with the same name from the same organization

### PR DESCRIPTION
**Pull request by rGregnanin**
Added the functionality that prevents the same Organization from creating two EServices with the same name

### CODE REVIEW

File: src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -13,7 +13,7 @@
         <appender-ref ref="CONSOLE" />
     </appender>
 
-    <root level="OFF">
+    <root level="ERROR">
```
Comment (bloodsky) [inline src/test/resources/logback-test.xml]: Revert here :leftwards_arrow_with_hook:

File: src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala [hunk 2]
```diff
@@ -67,6 +67,19 @@ final case class ProcessApiServiceImpl(
...
       clientSeed = Converter.convertToClientEServiceSeed(eServiceSeed, organizationId)
+      maybeEservice <- ReadModelQueries.listEServices(
+        eServiceSeed.name.some,
+        Seq.empty,
+        Seq(clientSeed.producerId.toString),
+        Seq.empty,
+        0,
+        1
+      )(readModel)
+      eServiceName = maybeEservice.results.map(_.name)
```
Comment (bloodsky) [inline src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala]: Maybe it's more consistent this way?
`maybeEServiceName = maybeEservice.results.headOption.map(_.name)`

File: src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala [hunk 3]
```diff
@@ -25,12 +25,15 @@ final case class CatalogManagementServiceImpl(invoker: CatalogManagementInvoker,
   implicit val logger: LoggerTakingImplicit[ContextFieldsToLog] =
     Logger.takingImplicit[ContextFieldsToLog](this.getClass)
 
-  override def createEService(eServiceSeed: EServiceSeed)(implicit contexts: Seq[(String, String)]): Future[EService] =
+  override def createEService(
+    eServiceSeed: EServiceSeed
+  )(implicit contexts: Seq[(String, String)]): Future[EService] = {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala]: We prefer avoiding brackets when not strictly required

File: src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala [hunk 4]
```diff
@@ -123,9 +136,9 @@ final case class ProcessApiServiceImpl(
         limit: Int
       ): Future[PaginatedResult[CatalogItem]] = {
 
-        if (apiAgreementStates.isEmpty)
+        if (apiAgreementStates.isEmpty) {
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala]: same for brackets

File: src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala [hunk 5]
```diff
@@ -67,6 +67,19 @@ final case class ProcessApiServiceImpl(
     val result: Future[EService] = for {
       organizationId <- getOrganizationIdFutureUUID(contexts)
       clientSeed = Converter.convertToClientEServiceSeed(eServiceSeed, organizationId)
+      maybeEservice <- ReadModelQueries.listEServices(
+        eServiceSeed.name.some,
+        Seq.empty,
+        Seq(clientSeed.producerId.toString),
+        Seq.empty,
+        0,
+        1
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala]: ⚠️ the filter by name in this query uses a regex that returns values that **contain** the given string.
It means that if we have the following eservices:
1. Tax Service 1
2. Tax Service 2
3. Tax Service
4. SomethinElse Service

and the requester wants to create `Tax Service`, this query returns just one of the first 3 values, because the request is with `limit = 1`, and you may not obtain the correct record.

Your approach is correct, but you need to make one of these changes:
1. retrieve all eservices containing that name, and then check if a service with that exact name exists
2. create a new filter in the `listEServices` query with an exact case insensitive match

(I suggest number 2)

**Review by galales — CHANGES_REQUESTED** [review]

File: src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala [hunk 6]
```diff
@@ -25,7 +25,9 @@ final case class CatalogManagementServiceImpl(invoker: CatalogManagementInvoker,
   implicit val logger: LoggerTakingImplicit[ContextFieldsToLog] =
     Logger.takingImplicit[ContextFieldsToLog](this.getClass)
 
-  override def createEService(eServiceSeed: EServiceSeed)(implicit contexts: Seq[(String, String)]): Future[EService] =
+  override def createEService(
+    eServiceSeed: EServiceSeed
+  )(implicit contexts: Seq[(String, String)]): Future[EService] =
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala]: As a general rule of thumb, don't reformat code if it's not strictly needed

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala [hunk 7]
```diff
@@ -73,10 +73,13 @@ object ReadModelQueries {
     producersIds: Seq[String],
     states: Seq[EServiceDescriptorState],
     offset: Int,
-    limit: Int
+    limit: Int,
+    exactMatchOnName: Boolean = false
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala:77]: I had an idea about this boolean parameter. We may abstract over the query that is used in the aggregation. WDYT @galales?

File: src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala [hunk 8]
```diff
@@ -67,6 +67,20 @@ final case class ProcessApiServiceImpl(
...
       organizationId <- getOrganizationIdFutureUUID(contexts)
       clientSeed = Converter.convertToClientEServiceSeed(eServiceSeed, organizationId)
+      maybeEservice <- ReadModelQueries.listEServices(
+        eServiceSeed.name.some,
+        Seq.empty,
+        Seq(clientSeed.producerId.toString),
+        Seq.empty,
+        0,
+        1,
+        true
```
Comment (TonioGela) [inline src/main/scala/it/pagopa/interop/catalogprocess/api/impl/ProcessApiServiceImpl.scala]: If we're picking this way, i.e. using a boolean parameter, let's at least name it
```suggestion
        exactMatchOnName = true
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala (hunk 7 above)
Comment (galales) [reply src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala:77]: not sure I got what you mean, can you provide an example?

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala [hunk 9]
```diff
@@ -73,10 +73,13 @@ object ReadModelQueries {
...
   )(readModel: ReadModelService)(implicit ec: ExecutionContext): Future[PaginatedResult[CatalogItem]] = {
 
-    val query = listEServicesFilters(name, eServicesIds, producersIds, states)
+    val query =
+      if (exactMatchOnName) listEServicesFilters(name, eServicesIds, producersIds, states, exactMatchOnName = true)
+      else listEServicesFilters(name, eServicesIds, producersIds, states)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala]: this is equivalent to
```suggestion
    val query = listEServicesFilters(name, eServicesIds, producersIds, states, exactMatchOnName)
```

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala [hunk 10]
```diff
@@ -121,7 +125,8 @@ object ReadModelQueries {
     val statesFilter       = mapToVarArgs(statesPartialFilter)(Filters.or)
     val eServicesIdsFilter = mapToVarArgs(eServicesIds.map(Filters.eq("data.id", _)))(Filters.or)
     val producersIdsFilter = mapToVarArgs(producersIds.map(Filters.eq("data.producerId", _)))(Filters.or)
-    val nameFilter         = name.map(Filters.regex("data.name", _, "i"))
+    val nameFilter         =
+      if (exactMatchOnName) name.map(Filters.eq("data.name", _)) else name.map(Filters.regex("data.name", _, "i"))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala]: in case of `exactMatchOnName == true` we lose the case insensitive match.
We want to avoid the creation of "Tax Service" along with "TaX SeRvIcE".

If you want to keep the boolean approach, the solution could be
```suggestion
      if (exactMatchOnName) name.map(n => Filters.regex("data.name", s"^$n$$", "i") else name.map(Filters.regex("data.name", _, "i"))
```
We still use the regex feature to apply case insensitive search, but the query now requires that there should not be any character before (with special character `^`) and after (with special character `$`) the given name.

Note that the double `$` is required as an escape to the string interpolation.

File: src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala (hunk 6 above)
Comment (galales) [reply src/main/scala/it/pagopa/interop/catalogprocess/service/impl/CatalogManagementServiceImpl.scala]: +1
if you reformat this code the change should disappear

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala (hunk 7 above)
Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala:77]: Passing a boolean parameter that has a default is a bad practice. My idea was to pass the main query object as an argument, wdyt? I have half an idea, I'll write it down better.

**Review by TonioGela — APPROVED** [review]

File: src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala (hunk 7 above)
Comment (galales) [reply src/main/scala/it/pagopa/interop/catalogprocess/common/readmodel/ReadModelQueries.scala:77]: I agree on the default value for the boolean.

If I get what you mean, I would keep the queries and their implementation inside this helper.
It improves maintainability, especially considering that part of the queries cannot be checked at compile time, at least we know where all the possible impacts on data change may be.
TL;DR single-responsibility principle

**Review by galales — APPROVED** [review]
